R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+              m3u = meat3u, m3l = meat3l))
+}
+
+#' @title (Internal) Design-based variance for models without covariance adjustment
+#' @param x a fitted \code{teeMod} model
+#' @details Calculate bread matrix for design-based variance estimate for 
+#'  \code{teeMod} models without covariance adjustment and 
+#'  without absorbed effects
+#' @return design-based variance estimate of the main treatment effect estimate
+#' @keywords internal
+.get_DB_wo_covadj_se <- function(x, ...){
+  if (x@absorbed_intercepts) {
+    stop("x should not have absorbed intercepts")
+  }
+  if (!is.null(x$call$offset)){
+    stop("x should not have covariance adjustment")
+  }
+  agg <- .aggregate_to_cluster(x)
+  data <- agg[[1]]
+  block <- agg[[2]]
+  
+  ws <- data$.w
+  yobs <- data$.wy  # observed ys by cluster
+  zobs <- data[, agg[[3]]]  # observed zs by cluster
+  bid <- data[, block] # block ids of clusters
+  
+  nbk <- design_table(design=x@Design, x="treatment",y="block")
+  # number of units by block and treatment, B by K
+  small_blocks <- apply(nbk, 1, min) == 1 # small block indicator
+  
+  nb <- rowSums(nbk)
+  rho <- c(sum(ws * yobs * (1-zobs)) / sum(ws * (1-zobs)),
+           sum(ws * yobs * zobs) / sum(ws * zobs))
+  
+  gammas <- (nbk[bid,1]*(1-zobs) + nbk[bid,2]*zobs) * ws # pseudo outcomes
+  gamsbk <- list()  # s^2_b,j, sample variance
+  for (k in 1:2){
+    indk <- zobs == (k-1)
+    gammas[indk] <- gammas[indk] * (yobs[indk] - rho[k])
+    gamsbk[[k]] <- aggregate(gammas[indk], by = list(data[indk,block]), FUN = var)
+  }
+  gamsbk <- merge(gamsbk[[1]], gamsbk[[2]], by = "Group.1")[,2:3]
+  gamsbk[is.na(gamsbk)] <- 0
+  gamsb <- aggregate(gammas, by = list(bid), FUN = var)[,2] # B vector
+  
+  nu1 <- rowSums(gamsbk / nbk)
+  nu2 <- 2 / nbk[,1] / nbk[,2] * choose(nb,2) * gamsb - 
+    (1/nbk[,1] + 1/nbk[,2]) * ((nbk[,1]-1) *gamsbk[,1] + (nbk[,2]-1) *gamsbk[,2])
+  varest <- (sum(nu1[!small_blocks]) + sum(nu2[small_blocks])) / sum(data$.w0)^2
+  return(as.matrix(varest))
+}
+
+#' @title (Internal) Aggregate weights and outcomes to cluster level
+#' @param x a fitted \code{teeMod} model
+#' @details aggregate individual weights and outcomes to cluster weighted sums
+#' @return a list of 
+#' - a data frame of cluster weights, outcomes, treatments, and block ids;
+#' - treatment id column name; 
+#' - block id column name
+#' @keywords internal
+.aggregate_to_cluster <- function(x, ...){
+  ws <- if (is.null(x$weights)) 1 else x$weights
+  data_temp <- x$call$data
+  name_y <- as.character(x$terms[[2]]) # the column of y
+  data_temp <- cbind(data_temp, .w = ws, .w0 = ws / ate(x@Design, data=data_temp),
+                     .wy = ws * data_temp[, name_y])
+  
+  design_obj <- x@Design
+  name_trt <- colnames(design_obj@structure)[design_obj@column_index == "t"]
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
```

Comment:
Can use `var_names(design_obj, x)`, where `x` is `"t"`, `"b"`, or `"u"`. Also, can this function handle the case where there's no blocking in the `Design` so `name_blk` is empty?

Reply:
Good point! I've revised accordingly in commit https://github.com/benbhansen-stats/propertee/pull/185/commits/6eb8927664674088f61b3323683e7d592a2de431